src/lib/IDBDAO.js

```diff
@@ -0,0 +1,377 @@
+//    data store.  Obviously this will get slow if you store large amounts
+//    of data in the database.
+//  */},
+  constants: {
+    /** Global cache of the current transaction reference. Only element 0 is used. */
+    __TXN__: [],
+  },
+
+  properties: [
+    {
+      name:  'of',
+      required: true
+    },
+    {
+      name:  'name',
+      label: 'Store Name',
+      expression: function() {
+        return this.of.id;
+      }
+    },
+    {
+      name: 'indicies',
+      factory: function() { return []; }
+    },
+    {
+      /** @internal */
+      name: 'propKeys_'
+    }
+  ],
+
+  methods: [
+
+    function init() {
+
+      var propKeys = this.propKeys_ = {};
+      var properties = this.of.getAxiomsByClass(foam.core.Property);
+      for ( var i = 0 ; i < properties.length ; i++ ) {
+        var prop = properties[i];
+        propKeys[prop.name] = prop.name;
+        if ( prop.shortName ) propKeys[prop.shortName] = prop.name;
+      }
+
+      //this.withDB = this.openDB;
+    },
+
+    function deserialize(json) {
+      return foam.json.parse(foam.json.parseString(json)); //TODO: serialization
+    },
+
+    function serialize(obj) {
+      return foam.json.stringify(obj); //TODO: serialization
+    },
+
+    function openDB(cc) {
+      var indexedDB = window.indexedDB ||
+        window.webkitIndexedDB         ||
+        window.mozIndexedDB;
+
+      var request = indexedDB.open("FOAM:" + this.name, 1);
+      var self = this;
+
+      request.onupgradeneeded = (function(e) {
+        var store = e.target.result.createObjectStore(self.name);
+        for ( var i = 0; i < self.indicies.length; i++ ) {
+          store.createIndex(self.indicies[i][0], self.indicies[i][0], { unique: self.indicies[i][1] });
+        }
+      })
+
+      request.onsuccess = (function(e) {
+        cc(e.target.result);
+      })
+
+      request.onerror = function (e) {
+        console.log('************** failure', e);
+      };
+    },
+
+    function withStore(mode, fn) {
+      if ( mode !== 'readwrite' ) return this.withStore_(mode, fn);
+
+      var self = this;
+
+      if ( ! this.q_ ) {
+        var q = [fn];
+        this.q_ = q;
+        this.async(function() {
+          self.withStore_(mode, function(store) {
+            // console.log('q length: ', q.length);
+            if ( self.q_ == q ) self.q_ = undefined;
+            for ( var i = 0 ; i < q.length ; i++ ) q[i](store);
+          });
+        }, this.X)();
+      } else {
+        this.q_.push(fn);
+        // Diminishing returns after 10000 per batch
+        if ( this.q_.length == 10000 ) this.q_ = undefined;
+      }
+    },
+
+    function withStore_(mode, fn) {
+      var self = this;
+      if ( self.__TXN__[0] ) {
+        try {
+          fn.call(self, self.__TXN__[0]);
+          return;
+        } catch (x) {
+          self.__TXN__[0] = undefined;
+        }
+      }
+      self.openDB(function (db) { // TODO: memoize like in foam1
```

Comment:
You could probably memoize this by making withDB a promise of the return value of openDB.

this.withDB = new Promise(function(resolve, reject) { this.openDB(resolve); };

Then this.withDB.then(function(db) { <db - work> });